### PR TITLE
[stablehlo] Add batching dims to `stablehlo.gather` and `stablehlo.scatter`

### PR DESCRIPTION
**Pull request by tomnatan30**
Add `operand_batching_dims` and `start_indices_batching_dims` attributes to `stablehlo.gather`. `operand_batching_dims` refers to the dimensions of the `operand` that are treated as batch. `start_indices_batching_dims` refers to the dimensions of the `start_indices` that are treated as batch. The corresponding dimension sizes must be equal. The semantics is equivalent to concatenating the outputs of the gather with each slices of `operand` and `start_indices`.

Similarly, add `input_batching_dims` and `scatter_indices_batching_dims` attributes to `stablehlo.scatter`. `input_batching_dims` refers to the dimensions of each tensor in `inputs` that are treated as batch. `scatter_indices_batching_dims` refers to the dimensions of the `scatter_indices` that are treated as batch.

See https://github.com/openxla/stablehlo/pull/2084 for more information

### CODE REVIEW

File: stablehlo/reference/Ops.cpp
Comment (tomnatan30) [inline stablehlo/reference/Ops.cpp:1682]: I am following the description I added in the specs: `full_batching_index[d_input] = ...`

but it would be easier to just iterate over `llvm::zip_equal(operandBatchingDims, startIndicesBatchingDims)`

similarly, it would be easier to iterate over `llvm::enumerate(startIndexMap)` when populating `fullStartIndex`

Alternatively, we could either use the fact that `operandBatchingDims` is sorted, and iterate `startIndicesBatchingDims` to avoid the need for llvm::find(operandBatchingDims, dOperand), or change the constraints so that `startIndicesBatchingDims` is the sorted one of the two, and use that instead of `batchIndex[dStart - (dStart < indexVectorDim ? 0 : 1)`.

wdyt?

File: docs/spec.md
Comment (tomnatan30) [inline docs/spec.md:3012]: wdyt about this pseudo code? it's similar to the `full_start_index` but the `[d_start - (d_start < index_vector_dim ? 0 : 1)]` notation is a bit bizzare, is there a nicer way to express this?

File: stablehlo/dialect/TypeInference.cpp
Comment (ghpvnist) [reply stablehlo/dialect/TypeInference.cpp:267]: After some time of reviewing, I realized that you just refactored this function out. This requires a few changes, but I think this is worth the effort since it makes the code more concise. Lmk if this is too much, I can make a follow up PR if needed.

File: stablehlo/reference/Ops.cpp
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:1681]: The interpreter is implemented keeping it in sync with the spec in mind, so let's keep this implementation.

File: stablehlo/tests/interpret/gather.mlir
Comment (tomnatan30) [reply stablehlo/tests/interpret/gather.mlir:63]: Yes it was :) and I verified my calculations with the evaluator as the change to it is quite trivial.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:1283]: I quite like this error message from scatter:
```
emitOptionalError(location, "offset_dims[", i,
                               "]: ", offsetDims[i], " is out of bounds for ",
                               "implied result rank ", resultRank);
```

so I am going to expose a method like you suggested that emits this phrasing.

File: stablehlo/reference/Ops.cpp
Comment (tomnatan30) [reply stablehlo/reference/Ops.cpp:1694]: SGTM

File: stablehlo/dialect/StablehloOps.td
Comment (tomnatan30) [reply stablehlo/dialect/StablehloOps.td]: Done.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp]: Done. Not sure why I did that :)

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp]: I went with the phrasing from scatter, and included the full name.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp]: done.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:1232]: I just followed the convension in gather, where we have one comment for both the sorted and unique if statements, since it's the same constraint.

Should I add above each if for both?

File: docs/images/spec/scatter.svg
Comment (tomnatan30) [reply docs/images/spec/scatter.svg:1]: Done.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp]: llvm::concat<const int64_t> doesn't return an ArrayRef, so this would require isUnique to accept a templated container, is that what you were thinking?

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp]: Done.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:1246]: Removes those and all checks of isa<RankedTensorType>

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:1384]: Done.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp]: Done.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:1318]: Done.

File: docs/spec.md
Comment (tomnatan30) [reply docs/spec.md:5084]: Done.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:1355]: Done.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:4415]: Done.

File: docs/spec.md
Comment (tomnatan30) [reply docs/spec.md:5034]: Done.

File: docs/spec.md
Comment (tomnatan30) [reply docs/spec.md:3103]: I only splitted it for cases where the is_unique check is done on multiple vectors, e.g. collapsed_slice_dims and operand_batch_dims. cases like offset_dims can stay as is since you still only care if they are unique locally. wdyt?

I can see why it would make sense to just always have is_unique and is_sorted separate, but can we do this in a follow up as this is quite a big change and it would be safer to make it as a single change?

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:1440]: See reply to related comment.

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [inline stablehlo/dialect/TypeInference.cpp:2560]: The message here is a bit confusing, do you have a better way to say that the slice size should be 1 because it's a collapsed/batching dim?

Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:2560]: Updated. wdyt?

**Review by GleasonK — APPROVED** [review]
Leaving my approval -- lgtm once @ghpvnist also approves!

File: stablehlo/dialect/StablehloOps.td
Comment (GleasonK) [inline stablehlo/dialect/StablehloOps.td:2716]: Let's talk about this change w.r.t. v1.0 -- I'm planning on making the cut on Monday, I think we should make this a post-1.0 change, since I think most users of StableHLO features have integrated with composites (0.19.0) and it's not great to immediately cut v1.0 after a fairly major op is updated

File: stablehlo/dialect/TypeInference.cpp
Comment (ghpvnist) [reply stablehlo/dialect/TypeInference.cpp:2619]: How about:
```
return emitOptionalError(location, "Expects that for each dim in ", name,
                         ", slice_sizes[dim] should be <=1, but got ",
                         sliceDimSize);
```

File: stablehlo/dialect/TypeInference.cpp
Comment (ghpvnist) [reply stablehlo/dialect/TypeInference.cpp]: The change looks great, thanks!

File: stablehlo/dialect/TypeInference.cpp
Comment (ghpvnist) [reply stablehlo/dialect/TypeInference.cpp]: You're right, then let's leave the current changes, and I'll revisit this later.

**Review by ghpvnist — CHANGES_REQUESTED** [review]
I left a few comments, rest looks good to me.

File: stablehlo/dialect/StablehloOps.td
Comment (tomnatan30) [reply stablehlo/dialect/StablehloOps.td:2744]: SGTM

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:2816]: SG!

File: stablehlo/transforms/CMakeLists.txt
Comment (tomnatan30) [reply stablehlo/transforms/CMakeLists.txt]: Done

File: stablehlo/dialect/TypeInference.cpp
Comment (tomnatan30) [reply stablehlo/dialect/TypeInference.cpp:2]: Done

File: BUILD.bazel
Comment (ghpvnist) [inline BUILD.bazel:319]: Remove extraneous line.

**Review by ghpvnist — APPROVED** [review]
LGTM. Thank you for the contribution!

File: BUILD.bazel
Comment (tomnatan30) [reply BUILD.bazel]: Done.

**Pull request merged**
